Fix `trainable_token_indices` for `lm_head`

aflueckiger opened the pull request:
It looks like that `trainable_token_indices` has been broken since #2605 for the `lm_head` in case the weights are not tied. The `lm_head` is an instance of `Linear` rather than `Embedding`, and, thus, it doesn't have an attribute `embedding_dim`.

The embedding dimension is not even needed further down unless when training with `deepspeed` or re-initializing the weights.

Asking @BenjaminBossan for a review due to the changes in #2605. I couldn't find any reports concerning that issue. Let me know if you would like to fix it differently.


Minimal config for reproduction:
```
model_name = "utter-project/EuroLLM-9B-Instruct"

trainable_tokens_indices = [5, 6, 7, 8, 9]

trainable_tokens = {"lm_head": trainable_tokens_indices, "embed_tokens": trainable_tokens_indices}

# Before this PR: Only training the embedding layer was possible
# trainable_tokens = {"embed_tokens": trainable_tokens_indices}
# trainable_tokens = trainable_tokens_indices

LoraConfig(
    lora_alpha=16,
    lora_dropout=0.05,
    r=16,
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=[
        "q_proj",
        "k_proj",
        "v_proj",
        "o_proj",
        "gate_proj",
        "up_proj",
        "down_proj",
    ],
    modules_to_save=None,
    trainable_token_indices=trainable_tokens,
)

```

BenjaminBossan reviewed (changes requested):
Thanks for providing this fix.

For the purpose of testing, could you please give a small example where this is broken? The model ID and PEFT config should be enough.

> The embedding dimension is not even needed further down unless when training with deepspeed or re-initializing the weights.

It's also used when `init_weights=False`. I guess we could move it into the corresponding branches to avoid retrieving this argument when it's not needed, but I think it's better to make the code robust enough that it does not break.

aflueckiger commented:
Thanks for the quick response. I added more information to reproduce the error in the PR description and replaced the try-expect block. Let me know about the preferred position for handling the attribute.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2863). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
